ClientRuntimes/Python/msrestazure/msrestazure/azure_active_directory.py

```diff
@@ -216,7 +228,70 @@ def clear_cached_token(self):
             raise_with_traceback(KeyError, "Unable to clear token.")
 
 
-class UserPassCredentials(OAuthTokenAuthentication, AADMixin):
+class AADRefreshMixin(object):
+    """
+    Additional token refresh logic
+    """
+
+    def refresh_session(self):
+        """Return updated session if token has expired, attempts to
+        refresh using newly acquired token.
+
+        :rtype: requests.Session.
+        """
+        if self.token.get('refresh_token'):
+            try:
+                return self.signed_session()
+            except Expired:
+                pass
+        self.set_token()
+        return self.signed_session()
+
+
+class AADTokenCredentials(AADMixin):
+    """
+    Credentials objects for AAD token retrieved through external process
+    e.g. Python ADAL lib.
+
+    Optional kwargs may include:
+    - china (bool): Configure auth for China-based service,
```

Comment:
rather than having a bool for China why not make the Crendentials class accept an AzureEnvironment object (Default is public azure).
From the environment object the credentials object will take the authentication endpoint. This will make it scalable as Azure keeps on adding more environments (USGovernment, BlackForest (German cloud), etc.). This also makes it easy for service teams to add a test environment like DogFood, etc. when they want to record tests against an internal envt. before going GA.
We did the same thing in node.js
- [AzureEnvironment](https://github.com/Azure/autorest/blob/master/ClientRuntimes/NodeJS/ms-rest-azure/lib/azureEnvironment.js)
- [UserTokenCredentials](https://github.com/Azure/autorest/blob/master/ClientRuntimes/NodeJS/ms-rest-azure/lib/credentials/userTokenCredentials.js#L22)

We also added wrapper methods (interactive, withUsernamePassword, withServicePrincipalSecret) with defaults to provide a similar experience like CLI. 
- [login.js](https://github.com/Azure/autorest/blob/master/ClientRuntimes/NodeJS/ms-rest-azure/lib/login.js)